src/bom/chainsop.rs

```diff
@@ -0,0 +1,872 @@
+                            .collect::<Vec<String>>()
+                            .join(",")
+                            ;
+                        args.push(OsString::from(bc_files))
+                    }
+                }
+        }
+        Ok(SubProcFile::NoOutputFile)
+    }
+
+    // Alternative to the setup_file function which instead uses the specified
+    // file instead of the existing specification, modifying the args in the
+    // appropriate manner to insert the file reference.
+    fn setup_file_override(&self,
+                           ovrf: &PathBuf,
+                           mut args: Vec<OsString>)
+                           -> Vec<OsString>
+    {
+        match &self {
+            FileSpec::Unneeded => args,
+            FileSpec::Append(_) => { args.push(ovrf.into()); args },
+            FileSpec::Option(flg, _) => {
+                args.push(OsString::from(flg));
+                args.push(ovrf.into());
+                args
+            },
+            FileSpec::Replace(pat, _) =>
+                args.into_iter()
+                .map(|arg| replace(pat, &ovrf.into(), &arg))
+                .collect()
+        }
+    }
+}
+
+
+// ----------------------------------------------------------------------
+// Single sub-process operation management
+
+/// This structure represents a single command to run as a sub-process, the
+/// command's arguments, and the input and output files for that sub-process.
+/// The structure itself is public but the fields are private
+/// (i.e. implementation specific); the impl section below defines the visible
+/// operations that can be performed on this structure.
+#[derive(Debug)]
+pub struct SubProcOperation {
+    cmd : Operation,
+    args : Vec<OsString>,
+    inp_file : FileSpec,
+    out_file : FileSpec,
+    in_dir : Option<PathBuf>,
+}
+
+enum Operation {
+    /// Name of executable to invoke in subprocess
+    Execute(OsString),
+
+    /// Local function to call instead of executing a subprocess.  The first
+    /// argument is the current working directory, the second is the argument
+    /// vector (the input and output files will be part of the argument vector as
+    /// determined by their corresponding FileSpec).
+    Call(Box<dyn Fn(&Path, Vec<OsString>) -> anyhow::Result<()>>)
+    // n.b. Would prefer this to be an FnOnce, but that breaks move semantics
+    // when trying to call it while it's a part of an enclosing Enum.
+}
+
+impl std::fmt::Debug for Operation {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result
+    {
+        match self {
+            Operation::Execute(cmd) => cmd.fmt(f),
+            Operation::Call(_) => "Local function call".fmt(f)
+        }
+    }
+}
+impl From<&Operation> for String {
+    fn from(op: &Operation) -> Self {
+        match op {
+            Operation::Execute(cmd) =>
+                cmd.clone().into_string().unwrap_or(String::from("<command>")),
```

Comment:
Can you just call `into_string` without `clone`? `into_string` should be making its own copy. (This is a question - I'm not sure).

Also this should probably be `.unwrap_or_else(|| String::from("<command>"))` so that the allocation of the string is lazy (only done in case of failure).

Reply:
The clone is required: without it `*cmd moved due to [into_string()] method call` and `move occurs because *cmd has type OsString, which does not implement the Copy constraint` and rustc recommends adding the `clone()`.

Made the error case lazy though, good idea.